feat(presence): Add support for signal batching

tylerbutler opened the pull request:
## Summary of changes

Presence updates are grouped together and throttled to prevent flooding the network with messages when presence values are rapidly updated. This means the presence infrastructure will not immediately broadcast updates but will broadcast them after a configurable delay.

The `allowableUpdateLatencyMs` property configures how long a local update may be delayed under normal circumstances, enabling batching with other updates. The default `allowableUpdateLatencyMs` is **60 milliseconds** but may be (1) specified during configuration of a States Workspace or Value Manager and/or (2) updated later using the `controls` member of Workspace or Value Manager. States Workspace configuration applies when a Value Manager does not have its own setting.

Notifications are never queued; they effectively always have an `allowableUpdateLatencyMs` of 0. However, they may be batched with other updates that were already queued.

Note that due to throttling, clients receiving updates may not see updates for all values set by another. For example,
with `Latest*ValueManagers`, the only value sent is the value at the time the outgoing batched message is sent. Previous
values set by the client will not be broadcast or seen by other clients.

### Example

You can configure the batching and throttling behavior using the `allowableUpdateLatencyMs` property as in the following example:

```ts
// Configure a states workspace
const stateWorkspace = presence.getStates("app:v1states",
	{
		// This value manager has an allowable latency of 100ms.
		position: Latest({ x: 0, y: 0 }, { allowableUpdateLatencyMs: 100 }),
		// This value manager uses the workspace default.
		count: Latest({ num: 0 }),
	},
	// Specify the default for all value managers in this workspace to 200ms,
	// overriding the default value of 60ms.
	{ allowableUpdateLatencyMs: 200 }
);

// Temporarily set count updates to send as soon as possible
const countState = stateWorkspace.props.count;
countState.controls.allowableUpdateLatencyMs = 0;
countState.local = { num: 5000 };

// Reset the update latency to the workspace default
countState.controls.allowableUpdateLatencyMs = undefined;
```

## Test cases

1. Signals send immediately when allowable latency is 0
1. Signals are batched when they are received within the send deadline (send deadline is set to now + allowable latency iff a new message is queued; i.e. a message was not previously queued - this is the first message to be put in the queue)
1. Signals from LVMs with different allowable latencies are batched correctly
1. Queued signal is sent with immediate signal (i.e. if a signal is queued and an immediate message comes in, it should merge and send immediately)
1. Multiple workspaces; messages from multiple workspaces should be batched and queued
1. Include sending notifications - they should be sent immediately

The test cases themselves include inline comments with the time and expected deadline for each operation. Signals are also numbered inline to make it easier to match up with the expected signals.

## Known gaps

LatestMapValueManager is not tested.

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
compiles and test passes

tylerbutler committed:
process incoming signal

tylerbutler committed:
sort of works

tylerbutler committed:
cleanup

tylerbutler committed:
updates

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Apply suggestions from code review

tylerbutler committed:
Apply suggestions from code review

tylerbutler committed:
revert changes in example

tylerbutler committed:
cleanup

tylerbutler committed:
simplify and correct tests

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
feedback

tylerbutler committed:
build

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
refactor(presence): Convert localUpdate to accept an options object

tylerbutler committed:
Merge branch 'main' into presence-batching

jason-ha committed:
test(client-presence): rework Notifications tests
- move workspace create to common setup
- add more test cases
- rework emit test without spies
- remove console logging
- remove compile coverage as part of regular tests

tylerbutler committed:
test cases for signals in latestValueManager

tylerbutler committed:
Merge pull request #30 from microsoft/presence/notification-events-tests

test(client-presence): rework Notifications tests

tylerbutler committed:
initial implemtnation

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
fix initial subscriptions not being triggered

tylerbutler committed:
Merge branch 'presence-notification-events' into presence-batching

tylerbutler committed:
Remove outdated comments

tylerbutler committed:
Remove outdated comments

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
wip

tylerbutler committed:
restore datastore manager

tylerbutler committed:
Revert "restore datastore manager"

This reverts commit eac63142a69a4b7e214f3229d13489b7bd79d7bb.

tylerbutler committed:
wip

tylerbutler committed:
vitest

tylerbutler committed:
snapshot tests

jason-ha commented on packages/framework/presence/src/test/latestValueManager.spec.ts:
```diff
@@ -3,12 +3,206 @@
+						},
+					},
+				],
+				[
+					"Pres:DatastoreUpdate",
+					{
+						"sendTimestamp": 1030,
+						"avgLatency": 10,
+						"data": {
+							"system:presence": {
+								"clientToSessionId": {
+									"client2": { "rev": 0, "timestamp": 1000, "value": "sessionId-2" },
+								},
+							},
+							"s:name:testStateWorkspace": {
+								"data": {
+									"sessionId-2": { "rev": 2, "timestamp": 1030, "value": { "num": 65 } },
+								},
+							},
+						},
+					},
+				],
```
If we drop the two processSignal calls, then I think this is the only signal expected when batching is implemented. The sendTimestamp should be 1070. (60 past when first local update was set at 1010.)
The test will also need a clock advance after last local set. I think you can clock.tick(1000) large number and timer would still happen at the right time.

tylerbutler committed:
updates

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
test updates

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -361,3 +411,35 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+	}
+
+	public clearTimeout(): void {
+		if (this.timeoutId !== null) {
```
```suggestion
		if (this.timeoutId !== undefined) {
```

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -361,3 +411,35 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+/**
+ * Timer manager to track whether timers have been triggered or not
+ */
+export class TimerManager {
```
```suggestion
class TimerManager {
```

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -361,3 +411,35 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 }
+
+/**
+ * Timer manager to track whether timers have been triggered or not
```
Wrapper around setTimeout to track whether the timeout has expired or not.

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +198,61 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+
+		const timeoutInMs = updateDeadline - now;
+		if (timeoutInMs > 0) {
+			this.timer.clearTimeout();
```
```suggestion
```

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +198,61 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+			// Queue the update
+			this.queuedMessage = newContent;
+			// if the timer has not expired, we can short-circuit because the timer will fire
+			// and cover this update
```
```suggestion
			// and cover this update. in other words, queuing this will be fast enough to
			// meet its deadline, because a timer is already scheduled to fire before its deadline.
```

tylerbutler committed:
Update presenceDatastoreManager.ts

tylerbutler committed:
Update presenceDatastoreManager.ts

tylerbutler committed:
Update presenceDatastoreManager.ts

tylerbutler committed:
Update presenceDatastoreManager.ts

tylerbutler committed:
Update presenceDatastoreManager.ts

tylerbutler committed:
clean up typing

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Merge branch 'presence-notification-events' into presence-batching

tylerbutler committed:
Merge branch 'presence-notification-events' into presence-batching

tylerbutler committed:
update tests

tylerbutler committed:
disable object sorting in snapshots

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
use ts-deepmerge

tylerbutler committed:
stringify snapshot since order controls don't seem to work

tylerbutler committed:
clarify that snapshot sort config doesn't work

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
add code coverage support

tylerbutler committed:
ignore test setup signals in snapshots

tylerbutler committed:
placeholder test

tylerbutler committed:
tests

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
new merge implementation

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler commented on packages/framework/presence/src/latestValueManager.ts:
```diff
@@ -102,7 +102,10 @@ class LatestValueManagerImpl<T, Key extends string>
 		this.value.value = value;
-		this.datastore.localUpdate(this.key, this.value, /* forceUpdate */ false);
+		this.datastore.localUpdate(this.key, this.value, {
+			forceBroadcast: false,
```
Should this be set to the controls.force value?

jason-ha replied:
There is no controls.force - we can remove froceBroadcast in favor of using allowableUpdateLatency from controls.

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -6,16 +6,24 @@
 import { assert } from "@fluidframework/core-utils/internal";
 import type { IInboundSignalMessage } from "@fluidframework/runtime-definitions/internal";
 import type { ITelemetryLoggerExt } from "@fluidframework/telemetry-utils/internal";
+// TODO: This lib weighs 585 bytes minified and gzipped; 1.1 kb minified only
+// https://bundlephobia.com/package/ts-deepmerge@7.0.1
+// If this is too heavy it should be possible to write a bespoke merger
```
```suggestion
```

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -366,3 +465,58 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+	}
+}
+
+// function deepMerge<T extends object>(target: T, source: T): T {
+// 	// eslint-disable-next-line no-restricted-syntax
+// 	for (const key in source) {
+// 		if (Object.prototype.hasOwnProperty.call(source, key)) {
+// 			if (isObject(source[key])) {
+// 				if (target[key] === undefined) {
+// 					// @ts-expect-error TODO
+// 					target[key] = {};
+// 				}
+// 				deepMerge(target[key] as T, source[key] as T);
+// 			} else {
+// 				target[key] = source[key];
+// 			}
+// 		}
+// 	}
+// 	return target;
+// }
+
+// function isObject(item: any): item is object {
+// 	// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/strict-boolean-expressions
+// 	return item && typeof item === "object" && !Array.isArray(item);
+// }
```
```suggestion

```

tylerbutler commented on packages/framework/presence/src/presenceStates.ts:
```diff
@@ -243,7 +259,7 @@ class PresenceStatesImpl<TSchema extends PresenceStatesSchema>
 
 			if (anyInitialValues) {
-				this.runtime.localUpdate(initial.newValues, false);
+				this.runtime.localUpdate(initial.newValues, { forceBroadcast: false });
```
I think this is the cause of the initial signals not getting queued. The localUpdate call here needs to pass the allowable latency. Not sure where to get it in this part of the code.

jason-ha replied:
Probably would want to accumulate from the value managers as initial.newValues is accumulated in the reduction above. We can just comment for now.

tylerbutler committed:
comments

tylerbutler committed:
policy

tylerbutler committed:
add back mocha test

tylerbutler committed:
Update packages/framework/presence/src/presenceDatastoreManager.ts

tylerbutler committed:
Update packages/framework/presence/src/presenceDatastoreManager.ts

tylerbutler committed:
lockfile

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +202,103 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+			this.sendMessageDeadline = updateDeadline;
+		}
+
+		// This function-local "datastore" will hold the merged message data
```
This is the core merge logic.

jason-ha commented on packages/framework/presence/package.json:
```diff
@@ -124,7 +126,8 @@
-		"@fluidframework/telemetry-utils": "workspace:~"
+		"@fluidframework/telemetry-utils": "workspace:~",
+		"ts-deepmerge": "^7.0.1"
 	},
```
remove - unused, right?

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +202,103 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+		const newMessage = {
+			sendTimestamp: now,
 			avgLatency: this.averageLatency,
 			// isComplete: false,
```
We'll want to queue just the data and add this stuff when we actually send. `now` will definitely change between then and now if queued. `averageLatency` might also get updated in the meantime.

tylerbutler replied:
Ahh, got it. I'll move this to the sendQueuedMessage function.

tylerbutler replied:
Indeed, moving this updated the snapshots with new send timestamps that correspond with the timestamp of the triggering event. Nice!

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +202,103 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+			// Since the key might not exist, create an empty object in that case. It will
+			// be set explicitly after the loop.
+			const mergedData = queueDatastore?.[workspaceName] ?? {};
+
```
Doesn't look like queueDatastore could be undefined here.
```suggestion
   const mergedData = queueDatastore[workspaceName] ?? {};
```

tylerbutler committed:
comments

tylerbutler committed:
move final message construction to immediately before send

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +202,103 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+				for (const [clientSessionId, value] of brandedObjectEntries(
+					workspaceData[valueManagerKey],
+				)) {
+					// TODO: Should any values be ignored here? E.g. values with ignoreUnmonitored?
```
For updates, each valueManager entry should have exactly one entry that is this client's session id... unless merging system workspace which is special...

See `localUpdate` on line 156. That guy does a lot of prep to form a full update message's content. Including prep of the system:presence workspace data that should be done last like sendTimestamp and averageLatency that I noted.

Probably I should have renamed the private function at some point and forceBroadcast argument should have been removed ...

The queued datastore updates should be handled in the caller.

Probably what we want is to have most of the content of line 156's localUpdate lambda be in method that sendQueuedMessage calls.
`queuedMessage` -> `queuedUpdateDatastore` (type `GeneralDatastoreMessageContent | undefined`)
From the new method remove the `updates` prep (that will be part of the merge you have now). `[internalWorkspaceAddress]: updates` will be replaced by `...queuedUpdateDatastore`.
```diff
   this.localUpdate(
    {
     // ...
     "system:presence": {
      clientToSessionId: {
       [clientConnectionId]: { ...currentClientToSessionValueState },
      },
     },
-     [internalWorkspaceAddress]: updates,
+     ...queuedUpdateDatastore,
    },
```

Then your merge is walking over more specific things: one workspace, one client session id for each value key.

That is probably far too much to handle in one read. Short is that code was not organized the best before (sorry) and the hook went in later than is ideal adding extra work and some complications. We should probably talk it through and hopefully the above text is good reference for making the changes.

jason-ha replied:
Update looks good enough for now

tylerbutler committed:
add targetClient to localUpdate options

tylerbutler committed:
Merge branch 'main' into presence-batching

jason-ha commented on packages/framework/presence/src/presenceStates.ts:
```diff
@@ -23,13 +23,21 @@ export type MapSchemaElement<
+	forceBroadcast: boolean;
+	allowableUpdateLatency?: number;
+}
+
```
Let's move this to stateDatastore.ts to avoid cyclical dependencies.
(I also think we can drop forceBroadcast in favor of allowableUpdateLatency to be 0 and mean the same thing.)

I have the API changes that I think make sense for a workspace to have default value. I think it make sense for me to create the same `LocalUpdateOptions` structure.

jason-ha replied:
As it turns out it makes sense for there to be two different options interfaces. So this can stay here.
I am naming this one RuntimeLocalUpdateOptions and the one for stateDatastores LocalStateUpdateOptions. ...but also not finalized so I reserved the right to change my mind. :)

jason-ha commented on packages/framework/presence/src/latestMapValueManager.ts:
```diff
@@ -191,7 +192,7 @@ class ValueMapImpl<T, K extends string | number> implements ValueMap<K, T> {
 			updates: InternalTypes.MapValueState<T>,
-			forceUpdate: boolean,
+			options: LocalUpdateOptions,
 		) => void,
```
I am eliminating this in my update. The value manager will have this. Shouldn't come from here. (bonus: eliminates a use of new type)

tylerbutler committed:
move generation of system workspace data to the send message function

tylerbutler committed:
Merge branch 'main' into presence-batching

jason-ha commented on packages/framework/presence/package.json:
```diff
@@ -79,12 +79,14 @@
 		"test:mocha:esm": "mocha --recursive \"lib/test/**/*.spec.*js\" --exit",
 		"test:mocha:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:mocha",
+		"test:vitest": "vitest run",
+		"test:vitest:coverage": "npm run test:vitest -- --coverage",
```
What does this actually test against? Is it doing ESM? How do I know?

tylerbutler replied:
It's the "include" setting in the vitest.config.ts file. Right now it's running against the src but we can switch to lib either in config or with a CLI flag. Do you want lib all the time or lib in CI, src locally?

jason-ha replied:
Well, I would say lib almost all of the time. It can run against src as a special local dev case.

tylerbutler replied:
It now runs against both src and lib, and outputs the snapshots for both versions of the test in the src/test/__snapshots__ folder. The tests run fast enough that running both isn't a big deal at the moment. I think we can further refine the config, but that can be done separately.

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +189,121 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+			return;
+		}
+
+		const clientConnectionId = this.runtime.clientId;
```
If not connected here skip the send. Should still clear the queued data.
```typescript
   // Check for connectivity before sending updates.
   if (this.runtime.connected) {
```

tylerbutler replied:
Won't clearing the queue result in messages being lost? Or is that what we want? Do we have a separate mechanism that handles disconnected messages? I guess we don't want the queue to grow infinitely...

jason-ha replied:
Notifications would be lost. States are kept in the datastore; so, when we reconnect all of those are broadcast as part of rejoin.

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +189,121 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+						mergedData[valueManagerKey][clientSessionId],
+						value,
+						0, // TODO: what value should be passed here?
+					);
```
0 is correct. This is a time shift for incoming updates. Nothing local needs a time shift.

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -195,14 +189,121 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+						0, // TODO: what value should be passed here?
+					);
+					console.log(clientSessionId, value);
+				}
```
remember to remove

tylerbutler committed:
run tests on lib

jason-ha commented on packages/framework/presence/src/test/mockEphemeralRuntime.ts:
```diff
@@ -162,15 +162,16 @@ export class MockEphemeralRuntime implements IEphemeralRuntime {
+	public readonly submittedSignals: Parameters<IEphemeralRuntime["submitSignal"]>[] = [];
+
+	public submitSignal(...args: Parameters<IEphemeralRuntime["submitSignal"]>): void {
+		this.submittedSignals.push(args);
```
This syntax doesn't make it readonly?
I used the other syntax to make it clearly replaceable.

tylerbutler replied:
I can revert these changes. This was from an early draft where I didn't subclass this mock runtime and tried to change it directly. I'll revert - not needed for this PR.

tylerbutler committed:
debug cleanup

tylerbutler committed:
revert unneeded mockruntime changes

tylerbutler committed:
missed comment

tylerbutler committed:
cleanup

tylerbutler committed:
format

jason-ha committed:
improvement(client-presence): broadcast control defaults
- `BroadcastControls` replace `LatestValueControls`
- `BroadcastControls` maybe specified on `PresenceStates` thru new `controls` property as defaults for all value managers.
   - `PresenceNotifications` redeclared apart from `PresenceStates` to avoid `controls` exposure. Also yields better API documentation.
- `allowableUpdateLatencyMs` was renamed from `allowableUpdateLatency` to clarify units are milliseconds. Specifying this value currently has no effect.
- Unsupported `forcedRefreshInterval` has been removed until implementation is closer.

jason-ha committed:
docs(client-presence): default restore  pattern
and improved check test

jason-ha committed:
revert(client-presence): reserve lint fixup
for targeted PR

tylerbutler committed:
Merge branch 'presence/broadcast-controls-api' into presence-batching-with-infra-updates

tylerbutler committed:
vitest config

tylerbutler committed:
clean up tests

tylerbutler committed:
updates

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-batching-with-infra-updates

jason-ha committed:
fix(client-presence): restore initial state values

tylerbutler committed:
Merge branch 'presence/fix-initial-state-data' into presence-batching

tylerbutler committed:
test updates

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
cleanup

tylerbutler committed:
make autosnapshotting optional

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
Revert "wip"

This reverts commit 481701cd044316838b3cd220645d86c8e7beb261.

tylerbutler committed:
tests

tylerbutler commented on packages/framework/presence/src/notificationsManager.ts:
```diff
@@ -150,7 +150,7 @@ class NotificationsManagerImpl<
 				{
 					rev: 0,
-					timestamp: 0,
+					timestamp: Date.now(),
```
probably should make these changes separately.

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -93,6 +97,44 @@ export interface PresenceDatastoreManager {
+		queueDatastore[workspaceName] = mergedData;
+	}
+	return queueDatastore;
+}
 /**
```
```suggestion
}

/**
```

tylerbutler committed:
Update packages/framework/presence/src/presenceDatastoreManager.ts

tylerbutler committed:
cleanup

tylerbutler committed:
rm notifications timestamps

tylerbutler committed:
timer manager

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
add mocha test for batching

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
mocha timer tests

tylerbutler committed:
rm vitest

tylerbutler committed:
cleanup

jason-ha commented:
Missing closing `)` in description:
> Signals are batched when they are received within the send deadline (send deadline is set to now + allowable latency iff a new message is queued (i.e. a message was not previously queued - this is the first message to be put in the queue)

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -200,14 +233,95 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+		const thisMessageDeadline = now + allowableUpdateLatencyMs;
+
+		if (
+			// If the deadline for this message is later than the overall send deadline, then
+			// we may be able to exit early if a timer will take care of sending it.
+			thisMessageDeadline >= this.timer.expireTime &&
+			// If the timer has not expired, we can short-circuit because the timer will fire
+			// and cover this update. In other words, queuing this will be fast enough to
+			// meet its deadline, because a timer is already scheduled to fire before its deadline.
+			!this.timer.hasExpired()
```
Should we swap the order? It feels odd to check the expireTime on something that has expired.

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -200,14 +233,95 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 		} satisfies DatastoreUpdateMessage["content"];
-		this.runtime.submitSignal(datastoreUpdateMessageType, content);
+		this.queuedData = undefined;
+		this.timer.clearTimeout();
```
Let's clear the timer first. It avoids having to think about od cases where there is an exception early that is caught and then timer fires later.

tylerbutler replied:
"first" as in before the previous call? Or first thing in the sendQueuedData function?

jason-ha replied:
First thing in the function.

jason-ha reviewed (commented):
Looks pretty good. Some minor changes requested.
Please also add a change set and update the readme.

I don't think there is a test that confirms some of the broadcast controls interactions but we can do that after.

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -93,6 +97,45 @@ export interface PresenceDatastoreManager {
+
+	// Merge the current data with the existing data, if any exists
+	// Iterate over the current message data; individual items are workspaces
+	for (const workspaceName of Object.keys(newData)) {
+		const workspaceData = newData[workspaceName];
```
There was some point where I had to do it this way to get the right typing, and then I thought I had to do it
everywhere. :)

tylerbutler commented on packages/framework/presence/src/test/batching.spec.ts:
```diff
@@ -0,0 +1,868 @@
+				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
+					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
+					immediateUpdate: Latest({ num: 0 }, { allowableUpdateLatencyMs: 0 }),
+				}); // SIGNAL #1 - not queued because it contains a value manager with a latency of 0,
+				// so the initial data will be sent immediately
+
+				const { count, immediateUpdate } = stateWorkspace.props;
+
+				clock.tick(10); // Time is now 1020
+				count.local = { num: 12 }; // will be queued; deadline is set to 1120
+
+				clock.tick(80); // Time is now 1100
+				count.local = { num: 34 }; // will be queued; deadline remains 1120
+
+				clock.tick(10); // Time is now 1110
+				immediateUpdate.local = { num: 56 };
+				// SIGNAL #2
+				// This should cause the queued signals to be merged with this immediately-sent
+				// signal with the value from the last signal (num=34).
+
+				// It's necessary to tick the timer beyond the deadline so the timer will fire.
+				clock.tick(10); // Time is now 1250
+
+				// expect(runtime.submittedSignals).toHaveLength(2);
```
I think these are artifacts of the refactoring from vitest to mocha. I think I've normalized them all now; the intended
format is that the SIGNAL comment appears either directly before the event that triggers it, or for signals that are
triggered by a timer, they'll still be before the clock tick call that triggers the timer, but I added a space between
the comment and the clock tick to indicate that it's kind of a separate event in the "timeline".

tylerbutler commented on packages/framework/presence/src/test/batching.spec.ts:
```diff
@@ -0,0 +1,868 @@
+				count.local = { num: 12 }; // will be queued; deadline remains 1060
+
+				clock.tick(30); // Time is now 1050
+				count.local = { num: 34 }; // will be queued; deadline remains 1060
+
+				clock.tick(10); // Time is now 1060
+				note.local = { message: "final message" }; // will be queued; deadline remains 1060
+
+				// SIGNAL #1
+				// At time 1060, the deadline timer will fire and send a single signal with the value
+				// from the last signal (num=34, message="final message").
+
+				// It's necessary to tick the timer beyond the deadline so the timer will fire.
+				clock.tick(30); // Time is now 1080
+
+				// expect(runtime.submittedSignals).toHaveLength(1);
```
There should still only be one signal, even if two timers get scheduled. The first scheduled timer fires at time 1060,
which is also when the second message gets queued, so the second gets batched but then the timer sends it immediately.
The second timer should no-op when it fires. Is that worth checking?

jason-ha replied:
It would be worth checking that there are no unexpected updates in general.
But here you would see updates. The `SIGNAL #1` comment should be with `clock.tick(10); // Time is now 1060` (which is the deadline).
See that after that clock advance there is `note.local = { message: "final message" }` and that state does not appear in any sent signal. If the final clock tick was bigger it would be sent. The final clock ticks should generally be large numbers to make sure there isn't anything left around.

tylerbutler replied:
> See that after that clock advance there is note.local = { message: "final message" } and that state does not appear in any sent signal. If the final clock tick was bigger it would be sent. The final clock ticks should generally be large numbers to make sure there isn't anything left around.

facepalm... I completely missed that the expected signal had the wrong message... Maybe I got it mixed up with another signal in all my scrolling. Regardless, I'll clean this up.

tylerbutler committed:
feedback

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
readme and changeset

tylerbutler committed:
docs

jason-ha commented on packages/framework/presence/README.md:
````diff
@@ -128,9 +128,33 @@ Notifications API is partially implemented. All messages are always broadcast ev
+### Throttling/batching
 
-Throttling is not yet implemented. `BroadcastControls` exists in the API to provide control over throttling of value updates, but throttling is not yet implemented. It is recommended that `BroadcastControls.allowableUpdateLatencyMs` use is considered and specified to light up once support is added.
+Presence signals are batched together and throttled to prevent flooding the network with signals when presence values are rapidly updated.
+The presence infrastructure will not immediately send outgoing signals; rather, they will be batched with any other outgoing signals and queued until a later time.
+
+A presence value manager such as `LatestValueManager` has an `allowableUpdateLatencyMs` value that can be configured; this value controls the longest time a signal will be queued.
+This value can be controlled dynamically at runtime, so adjustments can be made based on runtime data.
+
+The default `allowableUpdateLatencyMs` is **60 milliseconds**.
+
+Note that a signal may be sent before the allowable latency if another signal is sent. Signals are combined as needed and sent immediately at the earliest deadline.
+
+Notifications workspaces' signals are not queued; they effectively always have an `allowableUpdateLatencyMs` of 0. However, they may be batched with other signals that were queued earlier.
+
+The details of how signals are batched together is typically irrelevant to users of the Presence APIs, because the relevant value managers and the presence manager deal with these complexities.
+
+#### Example
+
+You can configure the batching and throttling behavior using the `allowableUpdateLatencyMs` property as in the following example:
+
+```ts
+// Configure a state workspace
+const stateWorkspace = presence.getStates("name:testStateWorkspace", {
+	// This value manager has an allowable latency of 100ms, overriding the default value of 60ms.
+	count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
+});
+```
````
We should avoid saying "signals" in most customer docs. Our use of Signals is an implementation detail. We can say updates or messages. These comments also miss to important aspects of controls - workspace settings and adjusting post creation (though does note it is possible).

````suggestion
Presence updates are grouped together and throttled to prevent flooding the network with messages when presence values are rapidly updated. This means the presence infrastructure will not immediately broadcast updates but will broadcast after a configurable delay.

 `allowableUpdateLatencyMs` configures how long a local update may be delayed under normal circumstances, enabling batching with other updates. The default `allowableUpdateLatencyMs` is **60 milliseconds** but may be 1) specified during configuration of a [States Workspace](#states-workspace) or [Value Manager](#value-managers) and/or 2) updated later using the `controls` member of Workspace or Value Manager. [States Workspace](#states-workspace) configuration applies when a Value Manager does not have its own setting.

Notifications are never queued; they effectively always have an `allowableUpdateLatencyMs` of 0. However, they may be batched with other updates that were already queued.

Note that due to throttling, clients receiving updates may not see updates for all values set by another. For example, with `Latest*ValueManagers` only the value at the time of update broadcast is sent.

#### Example

You can configure the batching and throttling behavior using the `allowableUpdateLatencyMs` property as in the following example:

```ts
// Configure a states workspace
const stateWorkspace = presence.getStates("app:v1states",
	{
		// This value manager has an allowable latency of 100ms.
		position: Latest({ x: 0, y: 0 }, { allowableUpdateLatencyMs: 100 }),
		// This value manager uses workspace default.
		count: Latest({ num: 0 }),
	},
	// Specify default for all value managers in workspace to 200ms, overriding the default value of 60ms.
	{ allowableUpdateLatencyMs: 200 }
);

// Temporarily set count update to send as soon as possible
const countState = stateWorkspace.props.count;
countState.controls.allowableUpdateLatencyMs = 0;
countState.local = { num: 5000 };
countState.controls.allowableUpdateLatencyMs = undefined; // Back to workspace default
```
````

jason-ha commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -9,16 +9,21 @@ import type { ITelemetryLoggerExt } from "@fluidframework/telemetry-utils/intern
 import type { ClientConnectionId } from "./baseTypes.js";
 import type { BroadcastControlSettings } from "./broadcastControls.js";
-import type { IEphemeralRuntime } from "./internalTypes.js";
+import { brandedObjectEntries, type IEphemeralRuntime } from "./internalTypes.js";
```
In this package we've been keeping type imports separate from non-type.

jason-ha commented on packages/framework/presence/src/test/batching.spec.ts:
```diff
@@ -0,0 +1,866 @@
+				const { note } = stateWorkspace2.props;
+
+				clock.tick(10); // Time is now 1020
+				note.local = { message: "will be queued" }; // will be queued, deadline is 1080
+				count.local = { num: 12 }; // will be queued; deadline remains 1080
+
+				clock.tick(30); // Time is now 1050
+				count.local = { num: 34 }; // will be queued; deadline remains 1080
+
+				clock.tick(10); // Time is now 1060
+				note.local = { message: "final message" }; // will be queued; deadline remains 1080
```
deadline in comments still says 1080 instead of 1070.

jason-ha reviewed (approved):
See above requests but looks ready to ship :)

Josmithr commented on .changeset/shy-files-jam.md:
```diff
@@ -0,0 +1,34 @@
+---
+"@fluidframework/presence": minor
```
I'm surprised not to see any API report updates - is this change not exported / user-facing? If not, we probably don't need a changeset entry. But if so, I would expect to see report changes.

tylerbutler replied:
There are no external API changes - but internally the behavior has changed. Messages were formerly sent immediately all the time; now they'll be batched/throttled by default.

Josmithr reviewed (approved):
Changeset looks good, though I did leave a question. Didn't review the code in detail, since that was already covered.

tylerbutler committed:
Update packages/framework/presence/README.md

Co-authored-by: Jason Hartman <jasonha@microsoft.com>

tylerbutler committed:
changeset

tylerbutler committed:
feedback

tylerbutler committed:
separate type imports

tylerbutler committed:
test cleanup

tylerbutler committed:
feedback

tylerbutler committed:
Merge branch 'main' into presence-batching

tylerbutler committed:
fix changeset links
